python/orca/src/bigdl/orca/learn/pytorch/pytorch_ray_worker.py

```diff
@@ -0,0 +1,123 @@
+
+
+import logging
+logger = logging.getLogger(__name__)
+
+try:
+    from collections.abc import Iterable
+except ImportError:
+    from collections import Iterable
+
+
+class PytorchRayWorker(TorchRunner):
+    """Manages a PyTorch model for training."""
+
+    def __init__(self,
+                 model_creator,
+                 optimizer_creator,
+                 loss_creator=None,
+                 metrics=None,
+                 scheduler_creator=None,
+                 training_operator_cls=None,
+                 config=None,
+                 use_tqdm=False,
+                 scheduler_step_freq=None):
+        super().__init__(model_creator, optimizer_creator, loss_creator, metrics, scheduler_creator,
+                         training_operator_cls, config, use_tqdm, scheduler_step_freq)
+
+        self.backend = "torch-local"
+        self.rank = 0
+        self.size = 0
+
+    def setup_horovod(self):
+        import horovod.torch as hvd
+        hvd.init()
+        self.backend = "horovod"
+        self.rank = hvd.rank()
+        self.size = hvd.size()
+        self.setup_components_horovod()
+        self.setup_operator(self.models)
+
+    def setup_address(self):
+        ip = self.get_node_ip()
+        port = find_free_port()
+        return f"tcp://{ip}:{port}"
+
+    def get_node_ip(self):
+        """Returns the IP address of the current node."""
+        return ray._private.services.get_node_ip_address()
+
+    def setup_components_horovod(self):
+        import horovod.torch as hvd
+
+        logger.debug("Creating model")
+        self.models = self.model_creator(self.config)
+        if not isinstance(self.models, Iterable):
+            self.models = [self.models]
+        else:
+            raise ValueError("only support single model for now")
+
+        assert all(isinstance(model, nn.Module) for model in self.models), (
+            "All models must be PyTorch models: {}.".format(self.models))
+
+        logger.debug("Creating optimizer.")
+        self.optimizers = self.optimizer_creator(self.given_models,
+                                                 self.config)
+        if not isinstance(self.optimizers, Iterable):
+            hvd.broadcast_parameters(self.models[0].state_dict(), root_rank=0)
+            hvd.broadcast_optimizer_state(self.optimizers, root_rank=0)
+            parameters = self.models[0].named_parameters()
+            self.optimizers = hvd.DistributedOptimizer(self.optimizers,
+                                                       named_parameters=parameters)
+            self.optimizers = [self.optimizers]
+        else:
+            raise ValueError("only support one optimizer for now")
+
+        self._create_schedulers_if_available()
+        self._create_loss()
+
+    def predict(self, data_creator, batch_size=32, profile=False):
+        """Evaluates the model on the validation data set."""
+        config = self.config.copy()
+        self._toggle_profiling(profile=profile)
```

Comment:
these two lines seem to be duplicated?

Reply:
Yeah... but I changed the predict data input for base class (`TorchRunner`) from `data creator` to `partition`, since that's where the common part starts for ray and pyspark backend.
And `config` will be used in `data_creator` to generate `partition`. The original code starts profiling before creating `partition`, and I kept consistent.